### DISCUSSION TITLE
3d probe and GRBL Hal/IO-Sender

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#241] [Q&A] [answered]
Hello,

I'm contemplating investing in a 3d probe, so i can get a more precise calibration of parts when i have to flip them over. 
Those looks very fancy, and has also got some fairly fancy pricetags, so I figure it would be wise to ask what and how it works with Grbl-hal and the io-sender. 

I do have a touchplate that i can put on top of the part, and I have tried using that for locating the precise placement of my part, but i find it rather difficult to get it to stay in place throughout the entire process, and then running an endmill up against the side of a stainless steel touchplate just makes my hairs stand on end... 

The plate i've got is this one:
https://makersupplies.dk/en/machin-plates/93-machine-plates-stainless-ox-zero-and-corner-finding-tool.html

I do see there is a good support for probing in IO Sender, so i should be able to connect a probe like this one:
https://www.ebay.co.uk/itm/CNC-3D-LOW-profile-digitizing-touch-probe-3axis-with-MAGNETIC-CONNECTOR/203197196915?_trkparms=aid%3D555023%26algo%3DHOMESPLICE.SIMRVI%26ao%3D1%26asc%3D20210125104017%26meid%3D39d72b8ad280423d9fbb0ac05ac8cc31%26pid%3D100752%26rk%3D1%26rkt%3D22%26mehot%3Dpf%26sd%3D251527393391%26itm%3D203197196915%26pmt%3D1%26noa%3D0%26pg%3D2047675%26algv%3DSimplAMLv5hPointwiseWebNoToraCoCoViewsNoHighIdfOrRoundRobinBlenderWithPromotedViewItems&_trksid=p2047675.c100752.m1982

or this one:
https://www.sorotec.de/shop/Usovo-3D-Messtaster.html

Would it be possible to connect these probes to the probe input (the first one seems possible, as it just has a VCC, GND and signal) - and then run a probing x, y and z and thereby find the precise location of a corner/hole of an object on the machine and use that to set G54 position?

Or am i completely barking up the wrong tree?

best
Kristoffer

**Comment by phil-barrett** [comment]
There are a number of people using 3D probes.  I use [this one.](http://www.craftycnc.com/touch-probes/). It is unpowered - ie just a switch which may have some longevity issues. Probe input needs to be inverted (it is NC - $6=1). The ebay link doesn't work so I'm not sure what is up with that - about 2 months ago the seller assured me that he was still selling them. He is not the most sophisticated seller on eBay. IIRC, it was $125+shipping. It works fine with center finding and height mapping in ioSender. I haven't tried TLO stuff, usually just make a height map, switch to my endmill/whatever, set Z with a touchplate and apply map to loaded GCode program.

The first probe you linked to (the 5-12V model) should work ok. I like the magnetic contact idea - in case of accidental spindle powering. I don't know if you need to invert probe with that - easy enough to figure out. The second link is a little light in detail with only 1 photo but says it is just a switch so it should work ok and uses inverted probe input. 

You are right about the price range of probes - there are some real nose bleed ones out there! They all seem to have center position adjustment.  I measured mine to about .05mm concentricity (is that the right word?) and could adjust better but that is the limit of my machine. The bottom end range of probes just have a switch while the better ones use electronics which allows them to reduce the contact current and, in theory, preserve contact life. Some of the better ones allow you to switch between NO and NC.

A point on NO vs NC. It is kind of a pain to mix 2 different probes - Touch plates are usually NO. Higher end CNC system have multiple probe inputs. I think this would be a good thing for grblHAL but not the end of the world. I have a "probe mixer" that I created. 2 probes can be connected, each can be set to NO or NC. The output goes into the probe input on grblHAL and can also be set NO or NC. LEDs to indicate which probe input is active and probe output state. Not sure if I want to make a product out of it. The current version uses an ATTINY but one made with just XOR gates would be easier to build.

**Reply by Klippert** [reply]
Hey Phil, 

Thanks for the nice explanation. My touch fixture is NC, which seems to work a bit more reliably than NO, as there are less noise issues. At least that was my experience with the touch plate - sometimes, if it touched some parts of the metal on the machine i would get triggering out of nowhere. 
It could be brilliant with a 3 input probe mixer. I.e. three inputs with either NC or NO and then convert that to a NC signal on the output that we feed back to the breakout board. This would allow one probe plate, one fixture and one spindle probe.  I'd love to have that flexibility. I think the input LED's would also be great, so one could easily see what's triggering.
If I remember my electronics right, that should be fairly simple using gates. 

these guys have also made one (https://www.sorotec.de/shop/Hot-plug-interface-for-Tool-lenght-sensor-and-3D-probe.html) that supports NPN and PNP inputs (which i remember has to do with transistors, but that's about as far as my recollection goes) as well as NO and NC inputs.  But i find it kind of expensive, and it requires joggling of cables when you switch probe. 

Anyhow - I'm happy to be your first customer if you make a design :-)

**Reply by phil-barrett** [reply]
The Sorotec product has a line that indicates it only combines the sensor signals: (translated) 

> You only have to pay attention that all sensors have the same switching function - either all Normally Closed or all Normally Open - so you do not have to constantly invert the function of the controller input.

I am a little surprised they did not add inversions. It would make the product a lot more flexible.

By the way, getting cables under control is a big issue.  Maybe more important than any fancy technology!

I decided not to turn it into a product because I did not see how to get it made at a price point that people would pay (<$50, my guess - what do you think is a reasonable price?).  A couple dollars worth of parts but way too many manufacturing steps including programming the ATTINY.  You can get Microchip to preprogram chips but then I would have to reflow them myself (exporting components to China is a big mess). USA production would price the product waaaay too high. A redesign that uses logic instead of code might get there. I am busy with other things right now but may look at it when I get some time.

**Reply by Klippert** [reply]
I think there is a decent market for something simple that allows us to use both probe fixtures and 3d probes/probe plates without having to faff with wiring and configuration. 
So count me in if you do go ahead with it :)

/K

**Reply by phil-barrett** [reply]
OK, you are now my product research!

Seriously, I will take a look.  I had originally thought just two inputs but 3 is clearly a requirement.  I'll have to look at how I would do that without a microcontroller. Maybe I just bite the bullet and design a programming station that takes a fully assembled board and automatically programs it.

A quick set of requirements:

- 3 inputs, switch to select
- NO and NC inputs supported
- 5V, 12V and 24V input supported
- 5V, 12V and 24V output supported
- input voltage autodetect
- NO or NC output supported
- LED input

feedback from anyone is appreciated.

**Reply by Klippert** [reply]
That sounds perfect to me.  Happy to be market research ;)

One question. some probes are also described as PNP/NPN probes. i get the feeling that works slightly different... aand i managed to answer my own question before finishing typing it up. just in case anyone else wants the explaination of NPN/PNP probes: https://www.se.com/uk/en/faqs/FA142566/

I'm almost tempted to say that it could be done entirely with gates? i.e. 
input -> opto coupler, then a switch (NO/NC) if the switch is set to NO signal is routed through a not gate, otherwise just straight to a mixer. i guess that woul make all signals NC, wchic means we could serial the inputs? 

I'm probably missing something, especially about the voltages, but it does seem that one could build something relatively inexpensive.

Oh, and if we could add a simple noise/jitter filter to the inputs, so that if someone touches a probe plate it doesn't go all over the place, that would be great. :-)


Best
Kristoffer

**Comment by einencool** [comment, marked as answer]
Hello Kristoffer,

I own the Probe from the ebay Link you mentioned above.
It‘s a really good product and worth its price.
A also have two „older“ Versions without the magnetic connection and use them for squaring my gantry, because they are absolutely reliable.

I‘m very happy with this Probe.
But you must tell him, if you want a NO or NC switch type. You can get different versions also for different voltage settings.
He‘s a really friendly and helpful seller.

Greets Chris

**Reply by Klippert** [reply]
Brilliant! Thanks einencool, it's always great to hear from someone who has hands on experience with the products.
I've already ordered one, and tracking says i'll have it on the 9th of april, so sitting tight... ;)
